### PR TITLE
Avoid duplicate exports of functions

### LINKED ISSUE #18095: Function manager exports duplicate functions
### What is the problem?

The function manager currently uploads functions, using a key derived from the pickled function, but couldpickle pickling isn't stable. This means the function will get reuploaded many times. 

On mitigation is to put a uuid in the function when the decorator is invoked, and use that (instead of hashing the bytes) to determine function uniqueness. 

*Ray version and other system information (Python version, TensorFlow version, OS):*

### Reproduction (REQUIRED)
Please provide a short code snippet (less than 50 lines if possible) that can be copy-pasted to reproduce the issue. The snippet should have **no external library dependencies** (i.e., use fake or mock data / environments):

```
@ray.remote
def f():
 pass

@ray.remote
def g():
  ray.get(f.remote())

ray.get([g.remote() for _ in range(5)])
```

(put a print statement in `function_manager.py::export`). 

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [ ] I have verified my script runs in a clean environment and reproduces the issue.
- [ ] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Don't export the same function multiple times since it's expensive: each export cause all workers to re-download the function.

## Related issue number
Closes #18095

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
Could you also try to run the original problematic case on a big cluster and check before/after performance/overheads is as expected?

File: python/ray/includes/function_descriptor.pxi [hunk 1]
```diff
@@ -189,11 +190,9 @@ cdef class PythonFunctionDescriptor(FunctionDescriptor):
         function_name = function.__qualname__
         class_name = ""
 
-        pickled_function_hash = hashlib.shake_128(pickled_function).hexdigest(
-          ray_constants.ID_SIZE)
-
         return cls(module_name, function_name, class_name,
-                   pickled_function_hash)
+                   hashlib.shake_128(uuid.bytes).hexdigest(
```
Comment (ericl) [inline python/ray/includes/function_descriptor.pxi:194]: Would uuid.uuid4().hex or something work?

File: python/ray/tests/test_basic_3.py [hunk 2]
```diff
@@ -291,5 +291,20 @@ def get_num_workers():
...
+    @ray.remote
+    def f():
+        pass
+
+    @ray.remote
+    def g():
+        ray.get(f.remote())
+
+    ray.get(g.remote())
+    num_exports = ray.worker.global_worker.redis_client.llen("Exports")
```
Comment (ericl) [inline python/ray/tests/test_basic_3.py:304]: Nice test!

File: python/ray/includes/function_descriptor.pxi (hunk 1 above)
Comment (jjyao) [reply python/ray/includes/function_descriptor.pxi]: Changed.

**Comment by jjyao** [comment]
> Could you also try to run the original problematic case on a big cluster and check before/after performance/overheads is as expected?

Checked before and after for #17970. Before it was slower and slower after each run and redis cpu was 100% all the time. Now the time is almost flat and redis cpu is low.

**Review by ericl — APPROVED** [review]

**Pull request merged**